### PR TITLE
fix(sandpack-react): safeguard against full page crash when there is no visible files

### PR DESCRIPTION
**Pull request by AaronPowell96**
## What kind of change does this pull request introduce?

Bux fix? 

TLDR: Added lots of optional chaining on "visibleFiles" and "code" to prevent errors where code is undefined or there are no visibleFiles.

This change prevents sandpack from erroring out in the scenario where there a no "active files" in view, which results in "code" being undefined among other things.

While this behaviour is not possible unless you start to play around with the 'deleteFile' consumer from context, I do believe  this would be a great "fix" to get in for people to play around further with customizing sandpack.
To clarify, this "bug" is not possible unless you use the sandpack provider and delete the active file with no other files being open.

## What is the current behavior?

image

Currently if you delete the last remaining active file, the whole page crashes.

## What is the new behavior?

New behaviour, the page doesn't crash and you receive a relevant error inside of the sandbox. This seems correct. Seen below (Deleted app.ts from the react.ts template). 
image

With the following console errors:
image

## What steps did you take to test this? This is required before we can merge, make sure to test the flow you've updated.

1. Create a local sandpack
2. Use the deleteFile from context on the only active file
3. Boom

## Checklist

- [ N/A ] Documentation;
- [ N/A ] Storybook (if applicable);
- [ N/A ] Tests;
- [X] Ready to be merged;

### CODE REVIEW

File: sandpack-react/src/components/CodeEditor/CodeMirror.tsx [hunk 1]
```diff
@@ -389,16 +389,23 @@ export const CodeMirror = React.forwardRef<CodeMirrorRef, CodeMirrorProps>(
...
-          ? EditorSelection.cursor(code.length)
-          : view.state.selection;
-
-        const changes = { from: 0, to: view.state.doc.length, insert: code };
-
-        view.dispatch({ changes, selection });
+        const selection =
+          view.state.selection.ranges.some(
+            ({ to, from }) => to > code?.length || from > code?.length
+          ) && code !== undefined
```
Comment (AaronPowell96) [inline sandpack-react/src/components/CodeEditor/CodeMirror.tsx:395]: This additional check means that when the only active file is deleted, there is no leftover code in the codeEditor.

File: sandpack-react/src/components/CodeEditor/CodeMirror.tsx [hunk 2]
```diff
@@ -389,16 +389,23 @@ export const CodeMirror = React.forwardRef<CodeMirrorRef, CodeMirrorProps>(
...
+            ? EditorSelection.cursor(code?.length)
+            : view.state.selection;
+
+        const changes = {
+          from: 0,
+          to: view.state.doc?.length,
+          insert: code ?? "",
```
Comment (AaronPowell96) [inline sandpack-react/src/components/CodeEditor/CodeMirror.tsx:402]: Inserting undefined created some errors with a CodeMirror plugin

**Review by danilowoz — COMMENTED** [review]
From the product point of view, I think it would be more interesting always show a file instead of an empty state. On delete, we could fall back to the next visible file, entry file, or the next file in the state. What do you think? 

That said, I think your changes in this PR are still valid, but I'm also interested in how it would look for the final users.

File: sandpack-react/src/components/CodeEditor/index.tsx [hunk 3]
```diff
@@ -102,7 +102,7 @@ export const SandpackCodeEditor = React.forwardRef<
     const { sandpack } = useSandpack();
     const { code, updateCode, readOnly: readOnlyFile } = useActiveCode();
     const { activeFile, status, editorState } = sandpack;
-    const shouldShowTabs = showTabs ?? sandpack.visibleFiles.length > 1;
+    const shouldShowTabs = showTabs ?? sandpack.visibleFiles?.length > 1;
```
Comment (danilowoz) [inline sandpack-react/src/components/CodeEditor/index.tsx:105]: When and why `visibleFiles` becomes undefined? It should always return an array.

**Comment by AaronPowell96** [comment]
> From the product point of view, I think it would be more interesting always show a file instead of an empty state. On delete, we could fall back to the next visible file, entry file, or the next file in the state. What do you think?
> 
> That said, I think your changes in this PR are still valid, but I'm also interested in how it would look for the final users.

I think this is an interesting approach too. When deleting the final activeFile having a codeEditor available with no code to edit / no files being changed did feel a little inconsistent.

I'll try to look into this today and raise an alternative PR.

File: sandpack-react/src/components/CodeEditor/CodeMirror.tsx (hunk 1 above)
Comment (AaronPowell96) [reply sandpack-react/src/components/CodeEditor/CodeMirror.tsx:395]: iirc I attempted this and it caused errors. I like your idea of falling back to the entry file and will attempt that first. Cheers for the feedback!

File: sandpack-react/src/components/CodeEditor/index.tsx (hunk 3 above)
Comment (AaronPowell96) [reply sandpack-react/src/components/CodeEditor/index.tsx:105]: > From the product point of view, I think it would be more interesting always show a file instead of an empty state. On delete, we could fall back to the next visible file, entry file, or the next file in the state. What do you think?
> 
> That said, I think your changes in this PR are still valid, but I'm also interested in how it would look for the final users.

I've reached out on twitter with a few questions, let me know if you'd prefer a github discussion etc!
